### PR TITLE
Support transactional systems (MicroOS)

### LINKED ISSUE #58519: [FEATURE REQUEST] Add support for transactional systems, like openSUSE MicroOS
Transactional systems, like MicroOS, provide a read-only rootfs and a cli tool (transactional-update) to manage the system (update, install new packages, etc).

Each change in this system become a snapper snapshot, that needs to be activated after a reboot.

Currently Salt do not support this kind of systems.

### PR DESCRIPTION
**Pull request by aplanas**
### What does this PR do?

Add support of transactional systems, like openSUSE MicroOS, into Salt.

This patch add two new execution modules, `transactional_update` and `rebootmgr`, and one execution module, `transactional_update`.

See the documentation for details about how it is used.

### Merge requirements satisfied?
- [X] Docs
- [X] Changelog
- [X] Tests written

Closes #58519

### CODE REVIEW

**Review by agraul — CHANGES_REQUESTED** [review]
Apart from the typos I found, this looks good to me.

I do have one question about something that is no obvious to me: why is `salt.client.ssh.state` used? Is it because the general approach is similar to salt-ssh (where needed things are copied over before executed) ?

File: salt/grains/extra.py [hunk 1]
```diff
@@ -66,3 +61,28 @@ def config():
...
+    grains = {
+        "efi": os.path.exists("/sys/firmware/efi/systab"),
+        "efi-secure-boot": __secure_boot(),
+    }
+
+    return grains
+
+
+def transactional():
+    """Determine if the system in transactional."""
```
[the remark this answers is not in the document]
Comment (aplanas) [reply salt/grains/extra.py:87]: thanks, done!

File: salt/modules/transactional_update.py [hunk 2]
```diff
@@ -0,0 +1,1270 @@
...
+
+Atomicity is the main feature of MicroOS, and to guarantee this
+property, this model leverages ``snapper``, ``zypper``, ``btrfs`` and
+``overlayfs`` to create snapshots that will be updated independently
+of the currently running system, and that are activated after the
+reboot.  This implies, for example, that some changes made on the
+system are not visible until the next reboot, as those changes are
+living in a different snapshot of the file system.
+
+This model present a lot of problems with the traditional Salt model,
```
[the remark this answers is not in the document]
Comment (aplanas) [reply salt/modules/transactional_update.py:22]: done!

File: salt/modules/transactional_update.py [hunk 3]
```diff
@@ -0,0 +1,1270 @@
...
+Continue a transaction
+......................
+
+One prerequisite already present is the support for branching from a
+different snapshot than the current one in snapper.
+
+With this feature we can represent in ``transactional-update`` the
+action of creating a transaction snapshot based on one that is planned
+to be the active one after the reboot.  This feature removes a lot of
+user complains (like, for example, loosing changes that are stored in
```
[the remark this answers is not in the document]
Comment (aplanas) [reply salt/modules/transactional_update.py:83]: Done!

**Comment by aplanas** [comment]
@dwoz no failing tests anymore

**Review by cmcmarrow — DISMISSED** [review]

**Comment by aplanas** [comment]
@dwoz @garethgreenaway ping? We should at least review the labels

File: salt/modules/transactional_update.py [hunk 4]
```diff
@@ -0,0 +1,1270 @@
...
+..................................................
+
+The option ``--drop-if-no-change`` is used to detect whether there is
+any change in the file system on the read-only subvolume of the new
+transaction will be added.  If a change is present, the new
+transaction will remain, if not it will be discarded.
+
+For example::
+
+  transactional-update --continute --drop-if-no-change run zypper in apache2"
```
Comment (agraul) [inline salt/modules/transactional_update.py:131]:
```suggestion
  transactional-update --continue --drop-if-no-change run zypper in apache2"
```
Just found this typo

Comment (aplanas) [reply salt/modules/transactional_update.py]: ouch, thanks! Fixed.

**Comment by sagetherage** [comment]
Since this is a new feature can you also add Release Notes, please?

**Review by garethgreenaway — DISMISSED** [review]

**Comment by aplanas** [comment]
> Release Notes

Sure! Is not https://github.com/saltstack/salt/pull/58520/files#diff-8f3bc23278640871a53d7ad4215e0f1ba92870fc85d132f051f9a12b224347fd what is for? I can extend this single line. Or there is a different process to add entries in the release notes?

**Comment by garethgreenaway** [comment]
@aplanas Those are the files that are generated for the changelog and that looks perfect, we were referring to the release notes which for the Aluminum release would be `doc/topics/releases/3003.rst`

**Comment by aplanas** [comment]
> @aplanas Those are the files that are generated for the changelog and that looks perfect, we were referring to the release notes which for the Aluminum release would be `doc/topics/releases/3003.rst`

Done in the last commit: https://github.com/saltstack/salt/pull/58520/commits/190a52f3db7ebfbdaa29d95ec66025dd7ad583b3

**Review by garethgreenaway — DISMISSED** [review]

**Comment by sagetherage** [comment]
this is looking like may need to defer this to another release

**Comment by aplanas** [comment]
> this is looking like may need to defer this to another release

@sagetherage What can I do to help here? The pytest fails are not from this PR, and was green months ago.

**Comment by sagetherage** [comment]
@aplanas  sure, there have been merges since months ago so there is expected merge conflicts and failing tests. Can you address the conflict, here, please? I will attempt to help get tests passing.

**Comment by aplanas** [comment]
> @aplanas sure, there have been merges since months ago so there is expected merge conflicts and failing tests. Can you address the conflict, here, please? I will attempt to help get tests passing.

Yes thanks! I just rebased all my commits, including this one.

**Comment by sagetherage** [comment]
It timed out so I am re-running that test. I will check back on that soon.

**Comment by sagetherage** [comment]
Tests are now passing, I will attempt to get it reviewed

**Comment by garethgreenaway** [comment]
@aplanas Can you move the release notes addition to the release notes for Silicon?  The 3003 release several months ago.  Thanks!

File: tests/unit/modules/test_rebootmgr.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
+import pytest
```
Comment (garethgreenaway) [inline tests/unit/modules/test_rebootmgr.py:1]: Can you make this a pytest test?

File: tests/unit/modules/test_systemd_service.py [hunk 6]
```diff
@@ -243,21 +243,27 @@ def test_available(self):
 
         # systemd < 231
         with patch.dict(systemd.__context__, {"salt.utils.systemd.version": 230}):
```
Comment (garethgreenaway) [inline tests/unit/modules/test_systemd_service.py:245]: Can you move this one to pytest?

File: tests/unit/modules/test_transactional_update.py [hunk 7]
```diff
@@ -0,0 +1,945 @@
+import sys
```
Comment (garethgreenaway) [inline tests/unit/modules/test_transactional_update.py:1]: Can you make this a pytest test?

File: tests/unit/modules/test_rebootmgr.py (hunk 5 above)
Comment (aplanas) [reply tests/unit/modules/test_rebootmgr.py:1]: Done.

File: tests/unit/modules/test_systemd_service.py (hunk 6 above)
Comment (aplanas) [reply tests/unit/modules/test_systemd_service.py:245]: This is an adjustment of the tests, not a new one, so I prefer for now leave it here. I will move all the new tests in this PR tho.

File: tests/unit/modules/test_transactional_update.py (hunk 7 above)
Comment (garethgreenaway) [reply tests/unit/modules/test_transactional_update.py:1]: Need to move this one over to pytest.  Thanks!

**Comment by garethgreenaway** [comment]
@aplanas Looks like there is a pre-commit failure for some black changes for one of the new modules.

File: tests/unit/modules/test_transactional_update.py (hunk 7 above)
Comment (aplanas) [reply tests/unit/modules/test_transactional_update.py]: Yes, done. Thanks!

**Review by garethgreenaway — DISMISSED** [review]

File: salt/utils/systemd.py [hunk 8]
```diff
@@ -51,6 +52,27 @@ def booted(context=None):
     return ret
 
 
+def offline(context=None):
+    """Return True is systemd is in offline mode"""
```
Comment (Ch3LL) [inline salt/utils/systemd.py:56]:
```suggestion
    """
    .. versionadded:: 3004
    
    Return True is systemd is in offline mode"""
```

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: doc/topics/releases/3004.rst [hunk 9]
```diff
@@ -12,3 +12,42 @@ Changed
...
+
+Atomicity is the main feature of MicroOS, and to guarantee this
+property, this model leverages ``snapper``, ``zypper``, ``btrfs`` and
+``overlayfs`` to create snapshots that will be updated independently
+of the currently running system, and that are activated after the
+reboot.  This implies, for example, that some changes made on the
+system are not visible until the next reboot, as those changes are
+living in a different snapshot of the file system.
+
+Salt 3003 (Aluminium) support this type of system via two new modules
```
[the remark this answers is not in the document]
Comment (aplanas) [reply doc/topics/releases/3004.rst:39]: Ouch. Done, thanks!

File: salt/utils/systemd.py (hunk 8 above)
Comment (aplanas) [reply salt/utils/systemd.py:56]: Done, but I added the versionadded after the description.

File: salt/modules/rebootmgr.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
+"""
+:maintainer:    Alberto Planas <aplanas@suse.com>
+:maturity:      new
+:depends:       None
+:platform:      Linux
```
[the remark this answers is not in the document]
Comment (aplanas) [reply salt/modules/rebootmgr.py:5]: Done!

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**